feat(client): add "completed x of y" in completion modal

Jagrutiti opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #46843 

<!-- Feel free to add any additional description of changes below this line -->

Jagrutiti committed:
feat: wrote a new function and updated the test case

ieahleen commented:
would you please follow the creating PR guidelines and complete the checklist items please?

Jagrutiti commented:
@ieahleen 

What exactly do you want me to do? I am missing code style or something?


 I do not think I am missing anything except one. I did not run the code myself because I don't know how to test this locally.

Kartik2220 commented:
you should check the boxes in boiler plate

Kartik2220 commented:
> @ieahleen
> 
> What exactly do you want me to do? I am missing code style or something?
> 
> I do not think I am missing anything except one. I did not run the code myself because I don't know how to test this locally.

you should check the boxes in boiler plate

Jagrutiti commented:
My bad @ieahleen I did not get you at first. 

How do I reproduce the error to test in local?

ieahleen commented:
@Jagrutiti to test your change, complete a Project in the New Responsive Web Design, the should say that you completed 1 of 5 projects, instead of seeing the 20%

majestic-owl448 committed:
Merge branch 'main' into feature/ui-specify-number-of-projects-from-full-projects

majestic-owl448 committed:
fix: lint errors

ieahleen commented:
I have run `npm run format` and committed the result to your branch. The lint tests should now pass.

Jagrutiti commented:
I have completed 8 steps but did not see the message. Where is it exactly displayed? @ieahleen 

![image](https://user-images.githubusercontent.com/28809492/181923510-aa5273e6-ba96-482b-be95-49b46e96648c.png)

ieahleen commented:
Can I suggest to understand the issue before creating a PR next time?

The modal in question appear when you complete a final project. For responsive web design, those are Tribute Page, Survey Form, Technical Documentation Page, Personal Portfolio Page, Product Landing Page

ieahleen commented on package.json:
```diff
@@ -150,7 +150,7 @@
     "babel-eslint": "10.1.0",
     "babel-plugin-transform-imports": "2.0.0",
-    "cross-env": "7.0.3",
+    "cross-env": "^7.0.3",
```
why have you committed this changes to `package.json`?

Jagrutiti replied:
I will undo it.

ieahleen commented:
I have built your PR locally, and there is no change, it still shows 20%, not '1 of 5':
![image](https://user-images.githubusercontent.com/26656284/181924208-09733de3-e7b6-426a-9be0-3b63c44f60ce.png)

Jagrutiti commented:
> I have built your PR locally, and there is no change, it still shows 20%, not '1 of 5': ![image](https://user-images.githubusercontent.com/26656284/181924208-09733de3-e7b6-426a-9be0-3b63c44f60ce.png)

Could you show me how you did you build it? I mean did you complete the whole project manually? 

 I would like to test what I missed.

ieahleen commented:
I built freeCodeCamp following the instructions in contributing guidelines.
To test, I just copied a passing project from somewhere and run the tests in the Survey Form page in the curriculum.

Jagrutiti commented:
Thank you @ieahleen 

I have got the issue. I am working on it.

Jagrutiti commented:
Hey @ieahleen 

I added a console statement in `get_completed_projects` in `completion-modal.tsx`. The required data is being displayed and printed at the right time. 

I am not able to display it in the modal. I am a noob in React would need help here.

![image](https://user-images.githubusercontent.com/28809492/181925897-20bb022b-8b49-4da8-9879-628d0a9eb5e9.png)


![image](https://user-images.githubusercontent.com/28809492/181925864-e07a8ba8-31f7-438c-a80b-372b00d307c6.png)

Jagrutiti committed:
fix: Fixed the changes not being reflected in the code. And undid unwanted commits from the last time

Jagrutiti committed:
fix: resolved merge conflicts

Jagrutiti commented on client/src/templates/Challenges/components/completion-modal-body.tsx:
```diff
@@ -113,6 +114,13 @@ export class CompletionModalBody extends PureComponent<
           </div>
+          <div>
+            <h3>
+              {t('learn.project-complete', {
```
Here I am not able to translate the output. It is printing `learn.project-complete` as is.

![image](https://user-images.githubusercontent.com/28809492/182011773-37ed7469-2c33-45e8-9105-d4f8d7e9dc0a.png)
###

Sboonny replied:
ya that happened, because a change was made before the function re-renders. so you got string instead of normal value

Jagrutiti commented:
I did `npm run clean` and build it again. It displays the required text here.

I could not complete other challenges to test (because few test cases failed from the code I got from Google)

Let me know if need other changes regarding this. 

![image](https://user-images.githubusercontent.com/28809492/182013225-89b99584-9d12-43af-a122-c7945af81682.png)

Sboonny commented on package-lock.json:
```diff
@@ -70,7 +70,7 @@
         "babel-eslint": "10.1.0",
         "babel-plugin-transform-imports": "2.0.0",
-        "cross-env": "7.0.3",
+        "cross-env": "^7.0.3",
```
```suggestion
        "cross-env": "7.0.3",
```

ieahleen replied:
this is till here

Jagrutiti replied:
Done

Sboonny commented on package-lock.json:
```diff
@@ -92,7 +92,7 @@
         "markdownlint": "0.26.1",
         "mock-fs": "5.1.2",
-        "npm-run-all": "4.1.5",
+        "npm-run-all": "^4.1.5",
```
```suggestion
        "npm-run-all": "4.1.5",
```

Sboonny commented on package-lock.json:
```diff
@@ -38602,8 +38602,9 @@
     },
     "node_modules/npm-run-all": {
       "version": "4.1.5",
+      "resolved": "https://registry.npmjs.org/npm-run-all/-/npm-run-all-4.1.5.tgz",
+      "integrity": "sha512-Oo82gJDAVcaMdi3nuoKFavkIHBRVqQ1qvMb+9LHk/cF4P6B2m8aP04hGf7oL6wZ9BuGwX1onlLhpuoofSyoQDQ==",
```
```suggestion
```

Sboonny commented on package-lock.json:
```diff
@@ -38602,8 +38602,9 @@
       "version": "4.1.5",
+      "resolved": "https://registry.npmjs.org/npm-run-all/-/npm-run-all-4.1.5.tgz",
+      "integrity": "sha512-Oo82gJDAVcaMdi3nuoKFavkIHBRVqQ1qvMb+9LHk/cF4P6B2m8aP04hGf7oL6wZ9BuGwX1onlLhpuoofSyoQDQ==",
       "dev": true,
```
```suggestion
      "dev": true,
      "license": "MIT",
```

Sboonny commented on package-lock.json:
```diff
@@ -80839,6 +80840,8 @@
     },
     "npm-run-all": {
       "version": "4.1.5",
+      "resolved": "https://registry.npmjs.org/npm-run-all/-/npm-run-all-4.1.5.tgz",
+      "integrity": "sha512-Oo82gJDAVcaMdi3nuoKFavkIHBRVqQ1qvMb+9LHk/cF4P6B2m8aP04hGf7oL6wZ9BuGwX1onlLhpuoofSyoQDQ==",
```
```suggestion
```

Sboonny reviewed (commented):
Is the current state of the lock file necessary?

Jagrutiti commented:
> Is the current state of the lock file necessary?

I did change them. I guess they were added while merging and I missed it.

Jagrutiti commented:
Hey @Sboonny 

For some reason, I am not able to commit your suggestions. Every time I try it says: Refresh and try again.


![image](https://user-images.githubusercontent.com/28809492/182029992-fd011c5c-cef5-471d-aca4-d4030f997637.png)

Sboonny commented:
@Jagrutiti i think it's GitHub server acting weird, they should sort it with time.

ieahleen commented:
hey @Jagrutiti are you still interested in working on this? please let us know, it has been 3 weeks since your last update.

Jagrutiti committed:
fix: removing unwanted updates to the package-lock.json

Jagrutiti commented:
Hey @ieahleen 

My apologies. I still wasn't able to commit to the suggestions so I have done them manually. If there is anything else pending from my side let me know.

ieahleen commented:
you have still changes to package-lock.json, could you please check again?

majestic-owl448 committed:
Merge branch 'main' into feature/ui-specify-number-of-projects-from-full-projects

Jagrutiti committed:
fix: Removing unwanted changes from package-lock.json

ieahleen commented on package-lock.json:
```diff
@@ -80870,6 +80869,8 @@
     },
     "npm-run-all": {
       "version": "4.1.5",
+      "resolved": "https://registry.npmjs.org/npm-run-all/-/npm-run-all-4.1.5.tgz",
+      "integrity": "sha512-Oo82gJDAVcaMdi3nuoKFavkIHBRVqQ1qvMb+9LHk/cF4P6B2m8aP04hGf7oL6wZ9BuGwX1onlLhpuoofSyoQDQ==",
```
```suggestion
      "version": "4.1.5",
```

ieahleen replied:
this change is also still here

ieahleen commented on package-lock.json:
```diff
@@ -38609,7 +38609,6 @@
     "node_modules/npm-run-all": {
       "version": "4.1.5",
       "dev": true,
-      "license": "MIT",
```
this change is also still here

Jagrutiti replied:
Hey @ieahleen 

My bad. For some reason I am unable to commit any suggestions in this PR. It is getting confusing. 

Let me know if there is anything else.

Jagrutiti committed:
fix: removing unwanted changes from package-lock.json

majestic-owl448 committed:
Merge branch 'main' into feature/ui-specify-number-of-projects-from-full-projects

ieahleen commented:
Have you had chance to continue working on this? The tests are still failing

Jagrutiti commented:
Hey @ieahleen 

My apologies. I am checking them right now.

Jagrutiti committed:
fix: updating snapshots to fix the test case

Jagrutiti commented:
Hey @ieahleen 

My test was failing here: 

  ```
    92 |     const { container } = render(<Profile {...notMyProfileProps} />);
      93 |
    > 94 |     expect(container).toMatchSnapshot();
```

At line no. 94. When I went to the `toMatchSnapshot()` function it redirects here `node_modules\@types\jest\index.d.ts`. Where I found the following information:

```
toMatchSnapshot(snapshotName?: string): R;
        /**
         * This ensures that a value matches the most recent snapshot with property matchers.
         * Instead of writing the snapshot value to a .snap file, it will be written into the source code automatically.
         * Check out [the Snapshot Testing guide](http://facebook.github.io/jest/docs/snapshot-testing.html) for more information.
         */
        // tslint:disable-next-line: no-unnecessary-generics
```
I went to the mentioned link and found the [update snapshots command](https://jestjs.io/docs/snapshot-testing#updating-snapshots).
 
`jest --updateSnapshot
`

When I ran this command I got the following output:

```
Snapshot Summary
 › 2 snapshots updated from 2 test suites.

Test Suites: 97 passed, 97 total
Tests:       635 passed, 635 total
Snapshots:   2 updated, 28 passed, 30 total
Time:        53.808 s
Ran all test suites.
```
The two snapshots where my tests were failing were updated.

Then when I ran `npm run test:source --u` all the test cases passed 

```
Test Suites: 97 passed, 97 total
Tests:       635 passed, 635 total
Snapshots:   30 passed, 30 total
Time:        45.613 s, estimated 48 s
Ran all test suites.
```
Adding the comment here because it took me a while to figure it out.

Jagrutiti committed:
Merge branch 'main' into feature/ui-specify-number-of-projects-from-full-projects

Jagrutiti commented:
Hey @ieahleen 

Even then 4 test cases are failing. They are not reproducible locally.

Their link that says learn more is broken. How do I fix them?

I get the following errors:

```
  console.warn
    redux-observable | WARNING: this middleware is already associated with a store. createEpicMiddleware should be called for every store.

    Learn more: https://goo.gl/2GQ7Da

      40 |     );
      41 |   } else {
    > 42 |     store = reduxCreateStore(
         |             ^
      43 |       rootReducer,
      44 |       composeEnhancers(applyMiddleware(sagaMiddleware, epicMiddleware))
      45 |     );

```

```
    Please update the following components: Dropdown

      10 |
      11 | function renderWithRedux(ui: JSX.Element) {
    > 12 |   return render(<Provider store={createStore()}>{ui}</Provider>);
         |          ^
      13 | }
      14 |
      15 | describe('<certification />', () => {
```

ieahleen commented:
The jest snapshot needs to be updated
https://jestjs.io/docs/%3Cmark%3Esnapshot%3C/mark%3E-testing#updating-%3Cmark%3Esnapshot%3C/mark%3Es

raisedadead committed:
Merge branch 'main' into feature/ui-specify-number-of-projects-from-full-projects

raisedadead committed:
chore: update snapshot

majestic-owl448 reviewed (dismissed):
It works, it passes the tests now. I say let's have it go in ASAP so that people are less confused by the percentages when completing projects!

![image](https://user-images.githubusercontent.com/26656284/196958307-529cd9aa-f2fa-4010-8e22-7e72adf713ef.png)

ShaunSHamilton commented on client/src/templates/Challenges/components/completion-modal.tsx:
```diff
@@ -92,6 +92,22 @@ export function getCompletedPercent(
+    return currentBlockIds.includes(id);
+  });
+
+  return `${completedChallengesInBlock.length} of ${currentBlockIds.length}`;
```
We cannot use a string including english words as part of a translation. Also, it makes more sense for this to be either a `{numerator, denominator}` or `"<NUMERATOR>/<DENOMINATOR>"`

Jagrutiti replied:
Hey @ShaunSHamilton 

If,

>We cannot use a string including english words as part of a translation.

then what shall I use for variable names?

I think I am missing some thing here.

ShaunSHamilton commented on client/src/templates/Challenges/components/completion-modal.tsx:
```diff
@@ -168,7 +186,18 @@ export class CompletionModalInner extends Component<
+
+    return {
+      downloadURL: newURL,
+      completedPercent: completedPercent,
```
```suggestion
      completedPercent,
```

ShaunSHamilton commented on client/src/templates/Challenges/components/completion-modal.tsx:
```diff
@@ -168,7 +186,18 @@ export class CompletionModalInner extends Component<
+    return {
+      downloadURL: newURL,
+      completedPercent: completedPercent,
+      completedProjects: completedProjects
```
```suggestion
      completedProjects
```

ShaunSHamilton commented on client/src/templates/Challenges/components/completion-modal.tsx:
```diff
@@ -92,6 +92,22 @@ export function getCompletedPercent(
+  currentBlockIds: string[] = [],
+  currentChallengeId: string
+): string {
+  completedChallengesIds = completedChallengesIds.includes(currentChallengeId)
```
Adding the `currentChallengeId` here steps away from the central state management. We should not confuse ourselves by calling `currentChallengeId` as part of `compeltedChallengesIds`. Instead, including the comment below, this logic should be simplified to count the number of completed **challenges** in the block, and add one if necessary.

raisedadead replied:
It's been a while, Do you mind adding code change suggestions or pushing a commit to keep this moving?

ShaunSHamilton commented on client/src/templates/Challenges/components/completion-modal.tsx:
```diff
@@ -216,6 +245,7 @@ export class CompletionModalInner extends Component<
     } = this.props;
 
     const { completedPercent } = this.state;
+    const { completedProjects } = this.state;
```
```suggestion
    const { completedPercent, completedProjects } = this.state;
```

ShaunSHamilton reviewed (dismissed):
Overall, the main issue is use of English within the translation string. Additionally, we should be able to simplify things elsewhere.

Jagrutiti committed:
Update client/src/templates/Challenges/components/completion-modal.tsx

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

Jagrutiti committed:
Update client/src/templates/Challenges/components/completion-modal.tsx

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

Jagrutiti committed:
Update client/src/templates/Challenges/components/completion-modal.tsx

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

Jagrutiti committed:
fix: removed hard coded english words from translation text

ShaunSHamilton committed:
fix: only show output when id is cert project

ShaunSHamilton committed:
fix: correctly count completed challenges

ShaunSHamilton commented on client/src/templates/Challenges/components/completion-modal.tsx:
```diff
@@ -77,21 +77,32 @@ export function getCompletedPercent(
+) {
+  const currentChallengeAlreadyCompleted =
+    completedChallengesIds.includes(currentChallengeId);
+  return (
+    completedChallengesIds.filter(challengeId =>
+      currentBlockChallengeIds.includes(challengeId)
+    ).length + +(!currentChallengeAlreadyCompleted && 1)
+  );
```
I have probably made things worse here, but my brain cannot come up with the simplified version of this.

Jagrutiti replied:
I did read your comment to make the code more readable for ease of maintenance.

But I did not touch it because I do not have complete knowledge of the code base.

And was not sure which places to ammends.

ShaunSHamilton replied:
Oliver has simplified this.

ShaunSHamilton committed:
fix: update test and snapshot

[skip ci]

ShaunSHamilton committed:
fix: at least one challenge is always completed

When the modal pops-up, there is always at least one challenge counted as "complete"

ojeytonwilliams committed:
refactor: slightly improve readability

raisedadead reviewed (approved):
Tested. Looks beautiful. 

Thanks a lot, @Jagrutiti @ShaunSHamilton and @ojeytonwilliams for the code, and @ieahleen, @Sboonny  for the feedback.
